plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -25,32 +26,84 @@
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.net.URI;
+import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static java.lang.Math.min;
 import static java.util.Objects.requireNonNull;
 
 public class FileSystemExchangeSource
         implements ExchangeSource
 {
-    private final FileSystemExchangeStorage exchangeStorage;
     @GuardedBy("this")
-    private final Iterator<URI> files;
+    private final List<CompletableFuture<Void>> readFutures;
     @GuardedBy("this")
-    private final Iterator<Optional<SecretKey>> secretKeys;
-
-    @GuardedBy("this")
-    private SliceInput sliceInput;
+    private final BlockingQueue<Slice> pendingSlices;
```

Comment:
The size of a single `Slice` here could be quite small (kilobytes). Allocating a new slice and placing it into a blocking queue may cause unnecessary allocations, memory copies and lock contention. We can try to move ahead and give the current approach a shot, but it feels like it is likely to be a bottleneck. 

How about the next approach:

1. Each reader has a fixed pre-allocated buffer. The size can be fixed to `16MB` (assuming this will be our maximum allowed page size). 
2. Each reader loads an entire `16MB` in a background, concurrently. When all 16MB is loaded it gives a signal to the main thread that the buffer is ready for consumption. 
3. When the buffer is consumed we ask the reader to re-fill the buffer.
4. When there's an incomplete page at the end we copy it to the beginning of the buffer and ask the reader to fill the remainder and so on.

Reply:
The proposed approach is essentially double buffering. However, I believe it's inevitable to allocate memory for Slices, this approach seems to only create more memory overhead.

Reply:
>  I believe it's inevitable to allocate memory for Slices

You don't have to allocate a new chunk of memory for every page. Instead you could create a view over a bigger pre-allocated chunk of memory.

Reply:
As discussed, `read()` can be called concurrently from different exchange operators, and there is no convenient way to signal that a buffer has been completely consumed, so it's inevitable to allocate memory for Slices today. As stated below, "parallel read" implementation will be a separate PR in the future.